diff -Nru debootstrap-0.2.45/functions debootstrap-0.2.45.fakechroot/functions
--- debootstrap-0.2.45/functions	2004-09-13 22:15:59 +0200
+++ debootstrap-0.2.45.fakechroot/functions	2005-02-05 20:46:19 +0100
@@ -564,9 +564,16 @@
   on_exit "umount $TARGET/dev/pts || true"
   on_exit "umount $TARGET/dev/shm || true"
   on_exit "umount $TARGET/proc/bus/usb || true"
-  on_exit "umount $TARGET/proc"
-  umount $TARGET/proc 2>/dev/null || true
-  in_target mount -t proc proc /proc
+  if [ "$FAKECHROOT" != "true" ]; then
+    on_exit "umount $TARGET/proc"
+    umount $TARGET/proc 2>/dev/null || true
+    in_target mount -t proc proc /proc
+  else
+    for i in cmdline cpuinfo devices filesystems loadavg meminfo misc modules partitions stat swaps uptime version; do
+        cat /proc/$i > $TARGET/proc/$i
+    done
+    mkdir $TARGET/proc/1
+  fi
 }
 
 setup_devices () {
@@ -574,7 +581,9 @@
     (cd "$TARGET"; zcat $DEVICES_TARGZ | tar -xf -)
   else
     if [ -e /dev/.devfsd ]; then
-      in_target mount -t devfs devfs /dev
+      if [ "$FAKECHROOT" != "true" ]; then
+        in_target mount -t devfs devfs /dev
+      fi
     else
       error 1 NODEVTGZ "no %s. cannot create devices" "$DEVICES_TARGZ"
     fi
@@ -631,3 +640,138 @@
   eval `echo EXIT_THING_${N_EXIT_THINGS}=\"$1\"`
   N_EXIT_THINGS="$(( $N_EXIT_THINGS + 1 ))"
 }
+
+################################################################### fake tools
+
+install_fake_tools () {
+  mv "$TARGET/sbin/ldconfig" "$TARGET/sbin/ldconfig.REAL"
+  echo \
+"#!/bin/sh
+echo
+echo \"Warning: Fake ldconfig called, doing nothing\"" > "$TARGET/sbin/ldconfig"
+  chmod 755 "$TARGET/sbin/ldconfig"
+
+  echo \
+"/sbin/ldconfig
+/sbin/ldconfig.REAL
+fakechroot" >> "$TARGET/var/lib/dpkg/diversions"
+
+  mv "$TARGET/usr/bin/ldd" "$TARGET/usr/bin/ldd.REAL"
+  cat << 'END' > "$TARGET/usr/bin/ldd"
+#!/usr/bin/perl
+
+# fakeldd
+#
+# Replacement for ldd with usage of objdump
+#
+# (c) 2003 Piotr Roszatycki <dexter@debian.org>, GPL
+
+
+my %libs = ();
+my @ld_library_path = qw(/usr/lib /lib);
+my $status = 0;
+my $dynamic = 0;
+
+sub ldso($) {
+    my ($lib) = @_;
+    my @files = ();
+    
+    foreach my $ld_path (@ld_library_path) {
+	-f "$ld_path/$lib" or next;
+	$libs{$lib} = "$ld_path/$lib";
+	push @files, "$ld_path/$lib";
+    }
+
+    objdump(@files);
+}
+
+sub objdump(@) {
+    my (@files) = @_;
+    my @libs = ();
+
+    foreach my $file (@files) {
+	open OBJDUMP, "objdump -p $file 2>/dev/null |";
+	while (my $line = <OBJDUMP>) {
+	    $line =~ s/^\s+//;
+	    my @f = split (/\s+/, $line);
+	    if ($f[0] eq "Dynamic") {
+		$dynamic = 1;
+	    }
+	    $f[0] eq "NEEDED" or next;
+	    if (! defined $libs{$f[1]}) {
+	        $libs{$f[1]} = undef;
+		push @libs, $f[1];
+	    }
+	}
+	close OBJDUMP;
+    }
+
+    foreach my $lib (@libs) {
+	ldso($lib);
+    }
+}
+
+
+if ($#ARGV < 0) {
+    print STDERR "fakeldd: missing file arguments\n";
+    exit 1;
+}
+
+while ($ARGV[0] =~ /^-/) {
+    my $arg = $ARGV[0];
+    shift @ARGV;
+    last if $arg eq "--";
+}
+
+open LD_SO_CONF, "/etc/ld.so.conf";
+while ($line = <LD_SO_CONF>) {
+    chomp $line;
+    unshift @ld_library_path, $line;
+}
+close LD_SO_CONF;
+
+unshift @ld_library_path, split(/:/, $ENV{LD_LIBRARY_PATH});
+
+foreach my $file (@ARGV) {
+    %libs = ();
+    $dynamic = 0;
+
+    if ($#ARGV > 0) {
+	print "$file:\n";
+    }
+
+    if (! -f $file) {
+	print STDERR "ldd: $file: No such file or directory\n";
+	$status = 1;
+	next;
+    }
+
+    objdump($file);
+    
+    if ($dynamic == 0) {
+	print "\tnot a dynamic executable\n";
+	$status = 1;
+    } elsif (scalar %libs eq "0") {
+	print "\tstatically linked\n";
+    }
+
+    foreach $lib (keys %libs) {
+	if ($libs{$lib}) {
+    	    printf "\t%s => %s (0x00000000)\n", $lib, $libs{$lib};
+	} else {
+	    printf "\t%s => not found\n", $lib;
+	}
+    }
+    
+}
+
+exit $status;
+END
+  chmod 755 "$TARGET/usr/bin/ldd"
+
+  echo \
+"/usr/bin/ldd
+/usr/bin/ldd.REAL
+fakechroot" >> "$TARGET/var/lib/dpkg/diversions"
+
+}
diff -Nru debootstrap-0.2.45/potato debootstrap-0.2.45.fakechroot/potato
--- debootstrap-0.2.45/potato	2004-09-13 22:15:59 +0200
+++ debootstrap-0.2.45.fakechroot/potato	2005-02-05 20:47:09 +0100
@@ -42,8 +42,10 @@
     setup_proc
     setup_devices
 
-    ln "$TARGET/sbin/ldconfig.new" "$TARGET/sbin/ldconfig"
-    in_target /sbin/ldconfig
+    if [ "$FAKECHROOT" != "true" ]; then
+        ln "$TARGET/sbin/ldconfig.new" "$TARGET/sbin/ldconfig"
+        in_target /sbin/ldconfig
+    fi
 
     x_feign_install () {
         local pkg=$1
@@ -79,6 +81,10 @@
         ln -sf perl-5.005 "$TARGET/usr/bin/perl"
     fi
 
+    if [ "$FAKECHROOT" = "true" ]; then
+        install_fake_tools
+    fi
+
     x_core_install base-files base-passwd ldso
     x_core_install dpkg
 
@@ -97,7 +103,9 @@
     cp "$TARGET/bin/true" "$TARGET/sbin/start-stop-daemon"
 
     setup_dselect_method apt
-    on_exit "in_target umount /dev/pts"
+    if [ "$FAKECHROOT" != "true" ]; then
+	on_exit "in_target umount /dev/pts"
+    fi
 
     in_target dpkg --configure --pending --force-configure-any --force-depends
 
diff -Nru debootstrap-0.2.45/sarge debootstrap-0.2.45.fakechroot/sarge
--- debootstrap-0.2.45/sarge	2005-01-21 06:21:54 +0100
+++ debootstrap-0.2.45.fakechroot/sarge	2005-02-05 20:46:58 +0100
@@ -118,10 +118,15 @@
         chown 0.0 "$TARGET/etc/fstab"; chmod 644 "$TARGET/etc/fstab"
     fi
 
-    setup_proc
+    if [ "$FAKECHROOT" != "true" ]; then
+        setup_proc
+    fi
+
     setup_devices
 
-    in_target /sbin/ldconfig
+    if [ "$FAKECHROOT" != "true" ]; then
+        in_target /sbin/ldconfig
+    fi
 
     p () {
 	baseprog="$(($baseprog + ${1:-1}))"
@@ -173,6 +178,10 @@
         ln -sf /usr/share/zoneinfo/UTC "$TARGET/etc/localtime"
     fi
 
+    if [ "$FAKECHROOT" = "true" ]; then
+        install_fake_tools
+    fi
+
     p; progress $baseprog $bases INSTBASE "Installing base system" #4
     x_core_install $LIBC6
 
diff -Nru debootstrap-0.2.45/sid debootstrap-0.2.45.fakechroot/sid
--- debootstrap-0.2.45/sid	2005-01-21 06:21:40 +0100
+++ debootstrap-0.2.45.fakechroot/sid	2005-02-05 20:46:09 +0100
@@ -7,6 +7,10 @@
 
     base="adduser apt apt-utils libdb4.2 at base-config aptitude libsigc++-1.2-5c102 bsdmainutils console-common console-tools libconsole console-data cpio cron dhcp-client ed exim4 exim4-base exim4-config exim4-daemon-light libgnutls11 libgcrypt11 libgpg-error0 libopencdk8 libtasn1-2 fdutils gettext-base groff-base ifupdown info klogd libssl0.9.7 liblzo1 zlib1g liblockfile1 libpcre3 libwrap0 logrotate mailx man-db libgdbm3 manpages nano net-tools netbase netkit-inetd iputils-ping nvi ppp pppconfig pppoe pppoeconf libpcap0.7 sysklogd tasksel libtextwrap1 tcpd telnet libtext-iconv-perl wget $additional" 
 
+    if [ "$FAKECHROOT" = "true" ]; then
+        required="$required fakeroot fakechroot binutils"
+    fi
+
     without_package () {
         echo "$2" | tr ' ' '\n' | grep -v "^$1$" | tr '\n' ' '
     }
@@ -118,10 +122,16 @@
         chown 0.0 "$TARGET/etc/fstab"; chmod 644 "$TARGET/etc/fstab"
     fi
 
-    setup_proc
+
+    if [ "$FAKECHROOT" != "true" ]; then
+        setup_proc
+    fi
+
     setup_devices
 
-    in_target /sbin/ldconfig
+    if [ "$FAKECHROOT" != "true" ]; then
+        in_target /sbin/ldconfig
+    fi
 
     p () {
 	baseprog="$(($baseprog + ${1:-1}))"
@@ -173,6 +183,10 @@
         ln -sf /usr/share/zoneinfo/UTC "$TARGET/etc/localtime"
     fi
 
+    if [ "$FAKECHROOT" = "true" ]; then
+        install_fake_tools
+    fi
+
     p; progress $baseprog $bases INSTBASE "Installing base system" #4
     x_core_install $LIBC6
 
diff -Nru debootstrap-0.2.45/slink debootstrap-0.2.45.fakechroot/slink
--- debootstrap-0.2.45/slink	2004-09-13 22:15:58 +0200
+++ debootstrap-0.2.45.fakechroot/slink	2005-02-05 20:46:53 +0100
@@ -34,8 +34,10 @@
 
     mv "$TARGET/usr/bin/perl.dist" "$TARGET/usr/bin/perl"
 
-    ln "$TARGET/sbin/ldconfig.new" "$TARGET/sbin/ldconfig"
-    in_target /sbin/ldconfig
+    if [ "$FAKECHROOT" != "true" ]; then
+        ln "$TARGET/sbin/ldconfig.new" "$TARGET/sbin/ldconfig"
+        in_target /sbin/ldconfig
+    fi
 
     x_feign_install () {
         local pkg=$1
@@ -65,6 +67,10 @@
     x_core_install base-files base-passwd ldso
     x_core_install dpkg
 
+    if [ "$FAKECHROOT" = "true" ]; then
+        install_fake_tools
+    fi
+
     ln -sf /usr/share/zoneinfo/UTC "$TARGET/etc/localtime"
     x_core_install libc6
 
diff -Nru debootstrap-0.2.45/woody debootstrap-0.2.45.fakechroot/woody
--- debootstrap-0.2.45/woody	2004-09-13 22:15:58 +0200
+++ debootstrap-0.2.45.fakechroot/woody	2005-02-05 20:47:03 +0100
@@ -97,10 +97,15 @@
         chown 0.0 "$TARGET/etc/fstab"; chmod 644 "$TARGET/etc/fstab"
     fi
 
-    setup_proc
+    if [ "$FAKECHROOT" != "true" ]; then
+        setup_proc
+    fi
+
     setup_devices
 
-    in_target /sbin/ldconfig
+    if [ "$FAKECHROOT" != "true" ]; then
+        in_target /sbin/ldconfig
+    fi
 
     p () {
 	baseprog="$(($baseprog + ${1:-1}))"
@@ -150,6 +155,10 @@
         ln -sf /usr/share/zoneinfo/UTC "$TARGET/etc/localtime"
     fi
 
+    if [ "$FAKECHROOT" = "true" ]; then
+        install_fake_tools
+    fi
+
     p; progress $baseprog $bases INSTBASE "Installing base system" #4
     x_core_install $LIBC6
 
